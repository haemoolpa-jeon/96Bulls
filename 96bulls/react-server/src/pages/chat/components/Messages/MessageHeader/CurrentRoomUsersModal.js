import React, { Component } from 'react';
import Modal from 'react-modal';
import PropTypes from 'prop-types';
import UserInfoItem from '../../common/UserInfoItem';

class CurrentRoomUsersModal extends Component {
  displayUsers = users => users.length > 0 && users.map(user => (
    <UserInfoItem
      user={user}
      onClick={() => this.inviteUser(user)}
      key={user.id}
    />
  ));

  render() {
    const customStyles = {
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        width: '35rem',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        border: 'none',
        background: '#eee',
        padding: '3rem',
        boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
      },
    };

    const { isOpen, closeModal, currentRoomUsers } = this.props;
    return (
      <Modal
        shouldCloseOnOverlayClick
        isOpen={isOpen}
        onRequestClose={closeModal}
        style={customStyles}
        ariaHideApp={false}
      >
        <h1>Current Users</h1>
        {this.displayUsers(currentRoomUsers)}
      </Modal>
    );
  }
}

CurrentRoomUsersModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  currentRoomUsers: PropTypes.array.isRequired,
};

export default CurrentRoomUsersModal;
