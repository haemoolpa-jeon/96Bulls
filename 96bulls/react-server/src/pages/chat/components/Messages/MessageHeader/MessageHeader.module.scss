@import "../../../styles/variable";

.message-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 2rem;
  height: 61px;
}
.room-name {
  color: $color-grey-dark;
  font-size: 18px;
  font-weight: bold;
}

.room-info {
  display: flex;
  flex-direction: column;
}

.user-count {
  text-align: left;
  font-size: 14px;
  color: $color-tertiary-dark;
  margin-top: .5rem;
  font-weight: bold;
  &:hover {
    cursor: pointer;
    opacity: 0.5;
  }
}

.icon {
  color: $color-primary-dark;
  font-size: 32px !important;
  &:hover {
    cursor: pointer;
    opacity: 0.5;
  }
}
