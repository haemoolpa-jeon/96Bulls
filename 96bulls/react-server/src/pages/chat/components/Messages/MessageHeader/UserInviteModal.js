import React, { Component } from 'react';
import Modal from 'react-modal';
import firebase from '../../../../../config/firebase';
import { connect } from 'react-redux';
import { setCurrentRoom } from '../../../actions';
import PropTypes from 'prop-types';
import UserInfoItem from '../../common/UserInfoItem';

export class UserInviteModal extends Component {
  state = {
    currentRoom: this.props.currentRoom,
    users: [],
    roomsRef: firebase.database().ref('rooms'),
    usersRef: firebase.database().ref('users'),
  };

  componentDidMount() {
    this.addListeners();
  }

  /**
   * 컴포넌트가 unmount될 때 이벤트 리스너를 제거합니다.
   */
  componentWillUnmount() {
    this.removeListeners();
  }

  addListeners = () => {
    this.addUserListener();
  };

  /**
   * 이벤트 리스너를 등록합니다.
   * DB에서 유저가 추가 될 때마다 실행 됩니다.
   */
  addUserListener = () => {
    const { usersRef } = this.state;
    const loadedUsers = [];
    usersRef.on('child_added', (snap) => {
      loadedUsers.push(snap.val());
      this.setState({ users: loadedUsers, loading: false });
    });
  };

  removeListeners = () => {
    this.state.usersRef.off();
    this.state.roomsRef.off();
  };

  /**
   * 유저를 초대합니다.
   * room과 user는 many to many 관계이기 때문에
   * room id 밑에는 유저 정보를
   * user id 밑에는 방 정보를 저장합니다.
   */
  inviteUser = (user) => {
    const {
      roomsRef, usersRef, currentRoom,
    } = this.state;
    const { closeModal } = this.props;
    const key = currentRoom.id;

    const invitedRoom = {
      id: currentRoom.id,
      name: currentRoom.name,
    };

    const invitedUser = {
      id: user.id,
      name: user.name,
      avatar: user.avatar,
    };

    roomsRef
      .child(key)
      .child('users')
      .child(user.id)
      .set(invitedUser)
      .then(() => {
        console.log('add user in room');
      })
      .catch((err) => {
        console.error(err);
      });

    usersRef
      .child(user.id)
      .child('rooms')
      .child(key)
      .set(invitedRoom)
      .then(() => {
        console.log('room add in user');
      })
      .catch((err) => {
        console.error(err);
      });
    closeModal();
  };

  /**
   * 현재 방에 들어와있는 유저를 제외하고 초대 가능한 유저만 보여줍니다.
   */
  displayUsers = (users) => {
    const currentRoomUserIds = this.props.currentRoomUsers.map(user => user.id);
    return users.length > 0 && users
      .filter(user => !currentRoomUserIds.includes(user.id)).map(user => (
        <UserInfoItem
          user={user}
          onClick={() => this.inviteUser(user)}
          key={user.id}
        />
      ));
  };

  render() {
    const customStyles = {
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        width: '35rem',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        border: 'none',
        background: '#eee',
        padding: '3rem',
        boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
      },
    };

    const { isOpen, closeModal } = this.props;
    const { users } = this.state;
    return (
      <Modal
        shouldCloseOnOverlayClick
        isOpen={isOpen}
        onRequestClose={closeModal}
        style={customStyles}
        ariaHideApp={false}
      >
        <h1>Invite</h1>
        {this.displayUsers(users)}
      </Modal>
    );
  }
}

UserInviteModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  currentRoom: PropTypes.object,
  currentRoomUsers: PropTypes.array,
};

export default connect(null, { setCurrentRoom })(UserInviteModal);
