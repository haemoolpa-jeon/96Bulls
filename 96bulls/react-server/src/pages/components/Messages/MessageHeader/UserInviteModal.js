import React, { Component } from 'react';
import Modal from 'react-modal';
import firebase from '../../../../config/firebase';
import { connect } from 'react-redux';
import { setCurrentRoom } from '../../../actions';
import PropTypes from 'prop-types';
import UserInfos from '../../common/UserInfos';

export class UserInviteModal extends Component {
  state = {
    currentRoom: this.props.currentRoom,
    users: [],
    roomsRef: firebase.database().ref('rooms'),
    usersRef: firebase.database().ref('users'),
  };

  componentDidMount() {
    this.addListeners();
  }

  //Remove Event listener when unmounted
  
  componentWillUnmount() {
    this.removeListeners();
  }

  addListeners = () => {
    this.addUserListener();
  };

  //Add Event listner that runs when user is added on DB
  addUserListener = () => {
    const { usersRef } = this.state;
    const loadedUsers = [];
    usersRef.on('child_added', (snap) => {
      loadedUsers.push(snap.val());
      this.setState({ users: loadedUsers, loading: false });
    });
  };

  removeListeners = () => {
    this.state.usersRef.off();
    this.state.roomsRef.off();
  };

  //Invite users to join the chat room
  inviteUser = (user) => {
    const {
      roomsRef, usersRef, currentRoom,
    } = this.state;
    const { closeModal } = this.props;
    const key = currentRoom.id;

    const invitedRoom = {
      id: currentRoom.id,
      name: currentRoom.name,
    };

    const invitedUser = {
      id: user.id,
      name: user.name,
      avatar: user.avatar,
    };

    roomsRef
      .child(key)
      .child('users')
      .child(user.id)
      .set(invitedUser)
      .then(() => {
        console.log('add user in room');
      })
      .catch((err) => {
        console.error(err);
      });

    usersRef
      .child(user.id)
      .child('rooms')
      .child(key)
      .set(invitedRoom)
      .then(() => {
        console.log('room add in user');
      })
      .catch((err) => {
        console.error(err);
      });
    closeModal();
  };

  //Make only available users to display
  displayUsers = (users) => {
    const currentRoomUserIds = this.props.currentRoomUsers.map(user => user.id);
    return users.length > 0 && users
      .filter(user => !currentRoomUserIds.includes(user.id)).map(user => (
        <UserInfos
          user={user}
          onClick={() => this.inviteUser(user)}
          key={user.id}
        />
      ));
  };

  render() {
    const customStyles = {
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        width: '35rem',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        border: 'none',
        background: '#eee',
        padding: '3rem',
        boxShadow: '0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)',
      },
    };

    const { isOpen, closeModal } = this.props;
    const { users } = this.state;
    return (
      <Modal
        shouldCloseOnOverlayClick
        isOpen={isOpen}
        onRequestClose={closeModal}
        style={customStyles}
        ariaHideApp={false}
      >
        <h1>Invite</h1>
        {this.displayUsers(users)}
      </Modal>
    );
  }
}

UserInviteModal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  currentRoom: PropTypes.object,
  currentRoomUsers: PropTypes.array,
};

export default connect(null, { setCurrentRoom })(UserInviteModal);
