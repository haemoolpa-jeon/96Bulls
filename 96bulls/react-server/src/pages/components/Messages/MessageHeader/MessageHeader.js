import React, { Component, Fragment } from 'react';
import Paper from '../../common/Paper';
import PropTypes from 'prop-types';
import styles from './MessageHeader.module.scss';
import UserInviteModal from './UserInviteModal';
import CurrentRoomUsersModal from './CurrentRoomUsersModal';
import firebase from '../../../../config/firebase';

class MessageHeader extends Component {
  state = {
    UserListModalIsOpen: false,
    CurrentRoomUsersIsOpen: false,
    roomsRef: firebase.database().ref('rooms'),
    currentRoom: this.props.currentRoom,
    currentRoomUsers: [],
  };

  componentDidMount() {
    this.addListeners();
  }

  /**
   * 컴포넌트가 unmount될 때 이벤트 리스너를 제거합니다.
   */
  componentWillUnmount() {
    this.removeListeners();
  }

  addListeners = () => {
    this.addRoomListener();
  };

  openCurrentRoomUsersModal = () => {
    this.setState({ CurrentRoomUsersIsOpen: true });
  };

  closeCurrentRoomUsersModal = () => {
    this.setState({ CurrentRoomUsersIsOpen: false });
  };

  openModal = () => {
    this.setState({ UserListModalIsOpen: true });
  };

  closeModal = () => {
    this.setState({ UserListModalIsOpen: false });
  };

  /**
   * 이벤트 리스너를 등록합니다.
   * DB에서 roodId의 유저가 추가 될 때마다 실행 됩니다.
   */
  addRoomListener = () => {
    const { roomsRef, currentRoom } = this.state;
    const currentRoomUsers = [];
    if (currentRoom) {
      roomsRef.child(currentRoom.id).child('users').on('child_added', (snap) => {
        currentRoomUsers.push(snap.val());
        this.setState({ currentRoomUsers });
      });
    }
  };

  removeListeners = () => {
    this.state.roomsRef.off();
  };

  render() {
    const { currentRoom } = this.props;
    const { UserListModalIsOpen, currentRoomUsers, CurrentRoomUsersIsOpen } = this.state;
    return (
      <Fragment>
        <Paper className={styles['message-header']}>
          <div className={styles['room-info']}>
            <h3 className={styles['room-name']}>{`# ${currentRoom ? currentRoom.name : 'Welcome to Open Chat'}`}</h3>
            <button
              type="button"
              className={styles['user-count']}
              onClick={this.openCurrentRoomUsersModal}
              disabled={!currentRoom}
            >
              {`Number of Users : ${currentRoomUsers.length}`}
            </button>
          </div>
          <button type="button" onClick={this.openModal} disabled={!currentRoom}>
            Invite user
          </button>
        </Paper>
        <UserInviteModal
          isOpen={UserListModalIsOpen}
          closeModal={this.closeModal}
          currentRoom={currentRoom}
          currentRoomUsers={currentRoomUsers}
        />
        <CurrentRoomUsersModal
          isOpen={CurrentRoomUsersIsOpen}
          closeModal={this.closeCurrentRoomUsersModal}
          currentRoomUsers={currentRoomUsers}
        />
      </Fragment>
    );
  }
}

MessageHeader.propTypes = {
  currentRoom: PropTypes.object,
};

export default MessageHeader;
