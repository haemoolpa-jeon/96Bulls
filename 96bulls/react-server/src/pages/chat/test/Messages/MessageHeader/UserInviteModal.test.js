import React from 'react';
import { UserInviteModal } from '../../../components/Messages/MessageHeader/UserInviteModal';
import {shallow} from 'enzyme/build';

describe('UserInviteModal', () => {
  let component = null;
  const currentRoomUsers = [{
    avatar: 'http://wwww.mock-image-url1.com',
    name: 'myp1',
    email: 'myp1@daum.net'
  },{
    avatar: 'http://wwww.mock-image-url2.com',
    name: 'myp2',
    email: 'myp2@daum.net'
  }];
  const currentRoom = {
    id: 'abcde',
    name: 'hello world',
  };

  const isOpen = false;
  const mockCloseModal = jest.fn(e => e);

  it('초기 렌더링이 문제없이 되야함', () => {
    component = shallow(<UserInviteModal
      currentRoomUsers={currentRoomUsers}
      closeModal={mockCloseModal}
      currentRoom={currentRoom}
      isOpen={isOpen}
    />);
  });

  it('초기 렌더링 스냅샷 일치함', () => {
    expect(component).toMatchSnapshot();
  });
});
