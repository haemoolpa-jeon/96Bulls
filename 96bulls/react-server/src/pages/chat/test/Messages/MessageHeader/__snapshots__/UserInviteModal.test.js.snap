// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInviteModal 초기 렌더링 스냅샷 일치함 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  isOpen={false}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
  style={
    Object {
      "content": Object {
        "background": "#eee",
        "border": "none",
        "bottom": "auto",
        "boxShadow": "0px 1px 5px 0px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14), 0px 3px 1px -2px rgba(0, 0, 0, 0.12)",
        "left": "50%",
        "marginRight": "-50%",
        "padding": "3rem",
        "right": "auto",
        "top": "50%",
        "transform": "translate(-50%, -50%)",
        "width": "35rem",
      },
    }
  }
>
  <h1>
    초대하기
  </h1>
</Modal>
`;
